src/trajectory/util.cpp

```diff
@@ -151,36 +153,38 @@ UniqueInterpolatedPtr concatenate(
 //==============================================================================
 double findTimeOfClosestStateOnTrajectory(
     const Trajectory& traj,
-    const Eigen::VectorXd& referenceState,
+    const statespace::StateSpace::State* referenceState,
+    double& distance,
     double timeStep)
 {
-  auto stateSpace = traj.getStateSpace();
-  const std::size_t configSize
-      = static_cast<std::size_t>(referenceState.size());
-  if (configSize != stateSpace->getDimension())
-    throw std::runtime_error("Dimension mismatch");
+  auto stateSpace
+      = std::dynamic_pointer_cast<const StateSpace>(traj.getStateSpace());
+  if (!stateSpace)
+    throw std::runtime_error("Failed to convert statespace");
 
   double findTime = traj.getStartTime();
   double minDist = std::numeric_limits<double>::max();
 
   const common::StepSequence sequence(
       timeStep, true, true, traj.getStartTime(), traj.getEndTime());
 
-  auto currState = stateSpace->createState();
-  Eigen::VectorXd currPos(stateSpace->getDimension());
+  auto metric = createDistanceMetric(stateSpace);
+
   for (const double currTime : sequence)
   {
+    auto currState = stateSpace->createState();
```

Comment:
Is there a reason to create this inside the loop?

Reply:
Don't think so, I'll try it outside.